pkg/proxy/util/utils.go

```diff
@@ -311,3 +322,44 @@ func EnsureSysctl(sysctl utilsysctl.Interface, name string, newVal int) error {
 	}
 	return nil
 }
+
+// DialContext is a dial function matching the signature of net.Dialer.DialContext.
+type DialContext = func(context.Context, string, string) (net.Conn, error)
+
+// FilteredDialOptions configures how a DialContext is wrapped by NewFilteredDialContext.
+type FilteredDialOptions struct {
+	// DialHostIPDenylist restricts hosts from being dialed. If this is nil, IsProxyableHostname is used.
+	DialHostCIDRDenylist []*net.IPNet
+}
+
+// NewFilteredDialContext returns a DialContext function that only allows connections to
+// proxyable addresses, as defined by IsProxyableHostname, in addition to the denylist in
+// FilteredDialOptions. If the wrapped DialContext is nil, all DNS and TCP settings use
+// golang default values.
+func NewFilteredDialContext(wrapped DialContext, opts *FilteredDialOptions) DialContext {
+	if wrapped == nil {
+		wrapped = http.DefaultTransport.(*http.Transport).DialContext
+	}
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		resp, err := net.DefaultResolver.LookupIPAddr(ctx, address)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(resp) == 0 {
+			return nil, ErrNoAddresses
+		}
+
+		for _, host := range resp {
+			if err := isProxyableIP(host.IP); err != nil{
```

Comment:
@jsafrane @gnufied would like your thoughts on this, if always disallowing localhost is going to break users.

Reply:
It will break some of my test deployments with local-up-cluster.sh. I am not sure about real users though.

Reply:
This is to mitigate the problem described here: https://groups.google.com/g/kubernetes-security-announce/c/kEK27tqqs30

How important is it for local-up-cluster.sh to work with localhost? Are there workarounds?

Reply:
Since we want to backport this fix to older releases, I would be more cautious about changing default behavior that we think has the potential to break someone. We try to minimize changes for patch releases (and offer a way to dsiable it), but we can consider changing things in a new minor release after a deprecation period.

Reply:
Yeah, that makes sense. On the other hand, secure-by-default is a good stance. If the only impacted scenarios are testing and similar, and it can be fixed with a flag, would that be enough? 

@jsafrane: if we could disable the IsProxyable check through the command-line flag, would that be enough?

Proposal: if --dial-host-cidr-denylist=none, then we do no filtering. If it's empty or contains CIDR ranges, then we do IsProxyable in addition to any ranges passed.

Reply:
`--dial-host-cidr-denylist=none`: LGTM
@humblec, AFAIK our production Gluster / Heketi never works on 127.0.0.1, it always run in a container that has  a real IP address, right?

Reply:
@jsafrane yeah, the heketi server is running in a pod with an IP, what SC carry for contacting heketi service is a heketi service url on top of this.

Reply:
@mattcary as @msau42 and @jsafrane suggested, having a fallback option with a knob/flag like `secure-be-default` should help. Maybe some scenarios/usages are unknown/missed from our radar and cause the breakage due to this change.

Reply:
+1 for `--dial-host-cidr-denylist=none`. Let's make that the default behavior for the flag in this PR so it's opt-in for the cherry-picks, and then open a new PR to make it opt out (not for cherry-picking).